lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb

```diff
@@ -116,12 +138,19 @@ def with_message(message)
           self
         end
 
+        def allow_nil(allow_nil = true)
+          @allow_nil = allow_nil
+          self
+        end
+
         def matches?(subject)
           super(subject)
           @expected_message ||= :blank
 
           if secure_password_being_validated?
             disallows_and_double_checks_value_of!(blank_value, @expected_message)
+          elsif @allow_nil
+            allows_value_of(nil)
```

Comment:
Right, so `validate_presence_of` typically accesses the attribute (or association) you specify, then calls `.present?` on it and checks that this returns true; if not, then it adds a validation error.

If you pass an `:allow_nil` to `validate_presence_of`, you're telling it, don't bother with running the validation if the association is nil.

So if the association is not nil, then it will check to make sure it isn't nil (this check will always pass). If the association is nil, then no check occurs.

So there's never a case in which `.present?` returns false, which means your validation isn't really doing anything -- your record will be valid regardless of whether the association is set or not.